tests/cycling-robot.cpp

```diff
@@ -0,0 +1,218 @@
+  {
+    auto constraint_models = typename dynamics::MultibodyConstraintFwdDynamicsTpl<double>::RigidConstraintModelVector();
+    auto const [placement, support] = height(t);
+
+    if (support)
+      constraint_models.emplace_back(makeConstraint(model)).joint2_placement = placement;
+
+    auto const stage_space = stage_space_t(model);
+    auto const proximal_settings = pinocchio::ProximalSettings(1e-9, 1e-10, 10);
+    auto actuation_matrix = Eigen::MatrixXd::Zero(model.nv, nu(model)).eval();
+    actuation_matrix.bottomRows(nu(model)).setIdentity();
+
+    auto const ode = dynamics::MultibodyConstraintFwdDynamicsTpl<double>(
+      stage_space, actuation_matrix, constraint_models, proximal_settings);
+    return dynamics::IntegratorSemiImplEulerTpl<double>(ode, dt);
+  }
+
+  auto makeStage(double t, double dt, pinocchio::Model const& model)
+  {
+    auto const rcost = makeCost(t, model);
+    auto const dynModel = makeDynamicsModel(t, dt, model);
+
+    return StageModelTpl<double>(rcost, dynModel);
+  }
+
+} // namespace
+
+BOOST_AUTO_TEST_CASE(test_cycling_robot) {
+
+  auto const model = SimpleModel();
+  auto data = pinocchio::Data(model);
+
+  auto const nq = model.nq;
+  auto const nv = model.nv;
+  auto const nx = nq + nv;
+  auto const nu = ::nu(model);
+  auto constexpr nsteps = 10;
+  auto constexpr dt = 0.02;
+
+  std::cout << "nq: " << nq << ", nv: " << nv << ", nx: " << nx << ", nu: " << nu << std::endl;
+
+  auto q = pinocchio::neutral(model);
+  auto vq = Eigen::VectorXd::Zero(nv).eval();
+  auto x = Eigen::VectorXd::Zero(nx).eval();
+  x << q, vq;
+
+  pinocchio::computeAllTerms(model, data, q, vq);
+  pinocchio::updateFramePlacements(model, data);
+
+  auto const term_cost = makeCost(0., model);
+  auto problem = TrajOptProblemTpl<double>(x, nu, stage_space_t(model), term_cost);
+
+  for (auto i = 0; i < nsteps; ++i)
+    problem.addStage(makeStage(0., dt, model));
+
+  auto constexpr TOL = 1e-7;
+  auto constexpr max_iters = 100u;
+  auto constexpr mu_init = 1e-8;
+  auto constexpr verbosity = QUIET;
+
+  auto ddp = SolverProxDDPTpl<double>(TOL, mu_init, max_iters, verbosity);
+  ddp.rollout_type_ = RolloutType::LINEAR;
+  ddp.sa_strategy_ = StepAcceptanceStrategy::LINESEARCH_NONMONOTONE;
+  ddp.filter_.beta_ = 1e-5;
+  ddp.force_initial_condition_ = true;
+  ddp.reg_min = 1e-6;
+  ddp.linear_solver_choice = LQSolverChoice::SERIAL;
+
+  ddp.setup(problem);
+
+  auto converged = ddp.run(problem);
+  BOOST_CHECK(converged);
+
+  for (auto t = 0.; t < 5.; t += dt)
+  {
+    auto const t0 = std::chrono::steady_clock::now();
+    auto const x = ddp.results_.xs[1];
+    q = x.head(nq);
+    vq = x.tail(nv);
+    pinocchio::computeAllTerms(model, data, q, vq);
+    pinocchio::updateFramePlacements(model, data);
+
+    auto const stage = makeStage(t, dt, model);
+    problem.replaceStageCircular(stage);
+    ddp.cycleProblem(problem, stage.createData());
+    problem.term_cost_ = makeCost(t + dt, model);
```

Comment:
One thing we can do is update `problem.term_cost_`. This would require downcasting the cost function back to `CostStack` - you can keep the downcasted reference or pointer somewhere - and extracting the cost component you want modified.
```cpp
CostStack &term_cost = dynamic_cast<CostStack &>(*problem.term_cost_);
auto *quad_cost = term_cost.getComponent<QuadraticResidualCostTpl<double>>("frame_fn");
// manipulate the 'quad_cost' pointer
```

The first cast is what we do in the helpers for `StageModel` and `CostStack` (the `getComponent<T>(const Key &)` helper) to extract the concrete types of the stored `polymorphic<T>` objects.

I'd prefer if `polymorphic<T>` had its own downcasting helper, but it is what it is.